proto/p4/config/v1/p4types.proto

```diff
@@ -292,4 +292,166 @@ message P4NewTypeSpec {
   repeated StructuredAnnotation structured_annotations = 4;
 }
 
+//////// Generic Data types
+
+// Instead of duplicating the type spec for these
+// every time the type is used, we include the type spec once in this
+// GenericTypeInfo message and refer to the types by name in the
+// GenericDataTypeSpec message. We also
+// support annotations for these type specs which can be useful, e.g. to
+// identify well-known headers (such as ipv4).
+message GenericTypeInfo {
+  map<string, GenericStructTypeSpec> structs = 1;
+  map<string, GenericListTypeSpec> lists = 2;
+  map<string, GenericEnumTypeSpec> enums = 3;
+  map<string, GenericSerializableEnumTypeSpec> serializable_enums = 4;
+  map<string, GenericNewTypeSpec> new_types = 5;
+}
+
+// Describes Generic Data types.
+message GenericDataTypeSpec {
+  oneof type_spec {
+    GenericBitstringLikeTypeSpec bitstring = 1;
+    GenericFloatType float = 2;
+    GenericBoolType bool = 3;
+    GenericNamedType struct = 4;
+    GenericNamedType list = 5;
```

Comment:
This may be a duplicate with some other comments Sayan or others may add to this PR:

These variants of collections were discussed verbally during the 2023-May-19 meeting:

+ list - ordered, duplicates allowed
+ sets - unordered, no duplicates allowed
+ bag/multiset/some-other-name - unordered, duplicates allowed

That leaves open a fourth possibility that I don't know any name by which it is commonly called:

+ name? - ordered, no duplicates allowed

Andy's personal preference: If all of the first three variants are desired, it seems better to give those different names, rather than having a single `list` type with one flag for ordered/unordered, and another flag for duplicates allowed/not-allowed.

Reply:
I like lists, sets and multisets.
Maybe we should think about unions in future as well.

Reply:
Ah, I have heard the 4th one called an "ordered set" before, where order is significant, and no duplicates are allowed.  At least in Python and a few other high level programming languages.